crates/air-lib/interpreter-data/src/interpreter_data/verification.rs

```diff
@@ -0,0 +1,284 @@
+        peer_id: String,
+    },
+
+    #[error(
+        "inconsistent CID multisets on merge for peer {peer_id:?}, prev: {larger_cids:?}, current: {smaller_cids:?}"
+    )]
+    MergeMismatch {
+        peer_id: String,
+        larger_cids: Vec<Box<str>>,
+        smaller_cids: Vec<Box<str>>,
+    },
+}
+
+/// An util for verificating particular data's signatures.
+pub struct DataVerifier<'data> {
+    // a map from peer_id to peer's info (public key, signature, CIDS)
+    grouped_cids: HashMap<Box<str>, PeerInfo<'data>>,
+    particle_id: &'data str,
+}
+
+impl<'data> DataVerifier<'data> {
+    // it can be further optimized if only required parts are passed
+    // SignatureStore is not used elsewhere
+    pub fn new(
+        data: &'data InterpreterData,
+        particle_id: &'data str,
+    ) -> Result<Self, DataVerifierError> {
+        // it contains signature too; if we try to add a value to a peer w/o signature, it is an immediate error
+        let mut grouped_cids: HashMap<Box<str>, PeerInfo<'data>> = data
+            .signatures
+            .iter()
+            .map(|(public_key, signature)| {
+                (
+                    public_key.to_peer_id().to_string().into(),
+                    PeerInfo::new(public_key, signature),
+                )
+            })
+            .collect();
+
+        // fill PeerInfo's `cids` field, checking for peer IDs without a key
+        collect_peers_cids_from_trace(&data.trace, &data.cid_info, &mut grouped_cids)?;
+
+        // sort cids for canonicalization
+        for peer_info in grouped_cids.values_mut() {
+            peer_info.cids.sort_unstable();
+        }
+
+        Ok(Self {
+            grouped_cids,
+            particle_id,
+        })
+    }
+
+    /// Verify each peers' signatures.
+    pub fn verify(&self) -> Result<(), DataVerifierError> {
+        for peer_info in self.grouped_cids.values() {
+            peer_info
+                .public_key
+                .verify(&peer_info.cids, self.particle_id, peer_info.signature)
+                .map_err(|error| DataVerifierError::SignatureMismatch {
+                    error: error.into(),
+                    cids: peer_info.cids.clone(),
+                    peer_id: peer_info.public_key.to_peer_id().to_string(),
+                })?;
+        }
+        Ok(())
+    }
+
+    /// For each peer, merge previous and current CID multisets by determining the largest set.
+    ///
+    /// This code uses an invariant: peer's multiset of produced CIDs is always a superset of
+    /// previous invocation's multiset:
+    ///
+    /// A_0 ⊆ A_1 ⊆ ... ⊆ A_n.
+    ///
+    /// So, the largest multiset is selected as the result of merging, the invariant is checked,
+    /// and a error is returned if it is violated.
+    ///
+    /// If the multisets are of same size, they have to be equal.
+    // TODO enforce merging only verified sets
+    pub fn merge(mut self, other: Self) -> Result<SignatureStore, DataVerifierError> {
```

Comment:
I suggest using `previous`, `new` instead of `our` and `other` to be consistent with the overall terminology.

Reply:
It doesn't matter here which is previous and which is new/current.

Reply:
I can add a comment that the order is not important and previous/current terminology is not used deliberately.